File: kernel/arch/x86_64-pc/sys/apic.c
/*
 * Copyright 2021 - 2023 Neptune
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <asm/asm.h>
#include <cpu/msr.h>
#include <cpu_features.h>
#include <cpuid.h>
#include <debug/debug.h>
#include <fw/acpi.h>
#include <fw/madt.h>
#include <io/mmio.h>
#include <klibc/time.h>
#include <mm/vmm.h>
#include <sys/apic.h>
#include <sys/hpet.h>
#include <sys/isr.h>
#include <sys/pic.h>
#include <sys/prcb.h>
#include <sys/timer.h>

static uintptr_t lapic_addr = 0;
static bool x2apic = false;
uint32_t tick_in_10ms = 0;

// Converts xAPIC MMIO offset into x2APIC MSR
static inline uint32_t reg_to_x2apic(uint32_t reg) {
	uint32_t x2apic_reg = 0;
	// MSR 831H is reserved; read/write operations cause general-protection
	// exceptions. The contents of the APIC register at MMIO offset 310H are
	// accessible in x2APIC mode through the MSR at address 830H
	// -- Intel SDM Volume 3A 10.12.1.2 Note 4
	if (reg == 0x310) {
		x2apic_reg = 0x30;
	} else {
		x2apic_reg = reg >> 4;
	}
	return x2apic_reg + 0x800;
}

uint32_t lapic_read(uint32_t reg) {
	if (x2apic) {
		return rdmsr(reg_to_x2apic(reg));
	}
	return mmind((void *)lapic_addr + MEM_PHYS_OFFSET + reg);
}

void lapic_write(uint32_t reg, uint32_t value) {
	if (x2apic) {
		wrmsr(reg_to_x2apic(reg), value);
	} else {
		mmoutd((void *)lapic_addr + MEM_PHYS_OFFSET + reg, value);
	}
}

static void lapic_set_nmi(uint8_t vec, uint8_t current_processor_id,
						  uint8_t processor_id, uint16_t flags, uint8_t lint) {
	// A value of 0xFF means all the processors
	if (processor_id != 0xFF) {
		if (current_processor_id != processor_id) {
			return;
		}
	}

	// Set to raise in vector number "vec" and set NMI flag
	uint32_t nmi = 0x400 | vec;

	// Set to active low if needed
	if (flags & 2) {
		nmi |= 1 << 13;
	}

	// Set to level triggered if needed
	if (flags & 8) {
		nmi |= 1 << 15;
	}

	// Use the proper LINT register
	if (lint == 0) {
		lapic_write(0x350, nmi);
	} else if (lint == 1) {
		lapic_write(0x360, nmi);
	}
}

uint8_t lapic_get_id(void) {
	return (uint8_t)(lapic_read(0x20) >> 24);
}

void lapic_init(uint8_t processor_id) {
	kprintf("LAPIC: Setting up LAPIC on Processor %u\n", processor_id);
	uint64_t apic_msr = rdmsr(0x1B);
	// Set APIC enable flag
	apic_msr |= 1 << 11;
	uint32_t a = 0, b = 0, c = 0, d = 0;
	if (__get_cpuid(1, &a, &b, &c, &d)) {
		if (c & CPUID_X2APIC) {
			x2apic = true;
			// Set x2APIC flag if support is detected
			apic_msr |= 1 << 10;
		}
	}
	wrmsr(0x1B, apic_msr);

	// Initialize local APIC
	lapic_write(0x80, 0);
	lapic_write(0xF0, lapic_read(0xF0) | 0x100);
	if (!x2apic) {
		lapic_write(0xE0, 0xF0000000);
		lapic_write(0xD0, lapic_read(0x20));
	}

	// Set NMIs according to the MADT
	for (int i = 0; i < madt_nmis.length; i++) {
		struct madt_nmi *nmi = madt_nmis.data[i];
		lapic_set_nmi(2, processor_id, nmi->processor, nmi->flags, nmi->lint);
	}

	// Set up APIC timer

	// Tell APIC timer to divide by 16
	lapic_write(0x3E0, 3);
	// Set timer init counter to -1
	lapic_write(0x380, 0xFFFFFFFF);

	timer_sleep(10);

	// Stop the APIC timer
	lapic_write(0x320, 0x10000);

	// How much the APIC timer ticked in 10ms
	tick_in_10ms = 0xFFFFFFFF - lapic_read(0x390);

	// Start timer as periodic on IRQ 0
	lapic_write(0x320, 32 | 0x20000);
	// With divider 16
	lapic_write(0x3E0, 3);
	lapic_write(0x380, tick_in_10ms / 10);
}

static uint32_t ioapic_read(uintptr_t ioapic_address, size_t reg) {
	mmoutd((void *)ioapic_address + MEM_PHYS_OFFSET, reg & 0xFF);
	return mmind((void *)ioapic_address + MEM_PHYS_OFFSET + 16);
}

static void ioapic_write(uintptr_t ioapic_address, size_t reg, uint32_t data) {
	mmoutd((void *)ioapic_address + MEM_PHYS_OFFSET, reg & 0xFF);
	mmoutd((void *)ioapic_address + MEM_PHYS_OFFSET + 16, data);
}

static uint32_t get_gsi_count(uintptr_t ioapic_address) {
	return (ioapic_read(ioapic_address, 1) & 0xFF0000) >> 16;
}

static struct madt_ioapic *get_ioapic_by_gsi(uint32_t gsi) {
	// Search through every I/O APIC to find its GSI
	for (int i = 0; i < madt_io_apics.length; i++) {
		struct madt_ioapic *ioapic = madt_io_apics.data[i];
		if (ioapic->gsib <= gsi &&
			ioapic->gsib + get_gsi_count(ioapic->addr) > gsi) {
			return ioapic;
		}
	}

	// Return NULL if none was found
	return NULL;
}

void ioapic_redirect_gsi(uint32_t gsi, uint8_t vec, uint16_t flags) {
	// Get I/O APIC address of the GSI
	size_t io_apic = get_ioapic_by_gsi(gsi)->addr;

	uint32_t low_index = 0x10 + (gsi - get_ioapic_by_gsi(gsi)->gsib) * 2;
	uint32_t high_index = low_index + 1;

	uint32_t high = ioapic_read(io_apic, high_index);

	// Set APIC ID
	high &= ~0xFF000000;
	high |= ioapic_read(io_apic, 0) << 24;
	ioapic_write(io_apic, high_index, high);

	uint32_t low = ioapic_read(io_apic, low_index);

	// Unmask the IRQ
	low &= ~(1 << 16);

	// Set to physical delivery mode
	low &= ~(1 << 11);

	// Set to fixed delivery mode
	low &= ~0x700;

	// Set delivery vector
	low &= ~0xFF;
	low |= vec;

	// Active high(0) or low(1)
	if (flags & 2) {
		low |= 1 << 13;
	}

	// Edge(0) or level(1) triggered
	if (flags & 8) {
		low |= 1 << 15;
	}

	ioapic_write(io_apic, low_index, low);
}

void ioapic_redirect_irq(uint32_t irq, uint8_t vect) {
	// Use ISO table to find flags and interrupt overrides
	for (int i = 0; i < madt_isos.length; i++) {
		if (madt_isos.data[i]->irq_source == irq) {
			ioapic_redirect_gsi(madt_isos.data[i]->gsi, vect,
								madt_isos.data[i]->flags);
			return;
		}
	}

	ioapic_redirect_gsi(irq, vect, 0);
}

void apic_send_ipi(uint32_t lapic_id, uint32_t flags) {
	if (x2apic) {
		// Write MSR directly, because lapic_write receives a 32-bit argument
		// Whilst in x2APIC, 0x830 is a 64-bit register
		wrmsr(0x830, ((uint64_t)lapic_id << 32) | flags);
	} else {
		lapic_write(0x310, (lapic_id << 24));
		lapic_write(0x300, flags);
	}
}

void apic_eoi(void) {
	// Writing any other value different than 0 may cause a #GP exception
	lapic_write(0xB0, 0);
}

void timer_stop_sched(void) {
	lapic_write(0x380, 0);
	lapic_write(0x320, (1 << 16));
}

void timer_sched_oneshot(uint8_t isr, uint32_t us) {
	timer_stop_sched();
	lapic_write(0x320, isr | 0x20000);
	lapic_write(0x3E0, 3);
	lapic_write(0x380, ((tick_in_10ms * (us / 1000))) / 10);
}

void apic_init(void) {
	pic_init();

	lapic_addr = acpi_get_lapic();
	lapic_init(madt_local_apics.data[0]->processor_id);
	ioapic_redirect_irq(0, 48);
}

File: src/arch/x86_64/apic.rs
use core::sync::atomic::{AtomicBool, AtomicU64, Ordering};

use super::acpi;
use super::acpi::AcpiHeader;
use super::asm::{cpuid, outb, rdmsr, wrmsr};
use crate::mm::virt::HHDM_OFFSET;

const MADT_TYPE_PROCESSOR_LOCAL_APIC: u8 = 0;
const MADT_TYPE_NON_MASKABLE_INTERRUPT: u8 = 4;
const MADT_TYPE_LOCAL_APIC_ADDRESS_OVERRIDE: u8 = 5;

#[repr(C, packed)]
pub struct MadtHeader {
    pub entry_type: u8,
    pub record_length: u8,
}

#[repr(C, packed)]
pub struct Madt {
    pub header: AcpiHeader,
    pub local_apic_address: u32,
    pub flags: u32,
    pub entry0: MadtHeader,
}

#[repr(C, packed)]
pub struct MadtProcessorLocalApic {
    pub header: MadtHeader,
    pub acpi_processor_id: u8,
    pub apic_id: u8,
    pub flags: u32,
}

#[repr(C, packed)]
pub struct MadtNmi {
    pub header: MadtHeader,
    pub acpi_processor_id: u8,
    pub flags: u16,
    pub lint: u8,
}

static MADT_PTR: AtomicU64 = AtomicU64::new(0);

const MAX_MADT_ENTRIES: usize = 256;
static mut MADT_LOCAL_APICS: [*const MadtProcessorLocalApic; MAX_MADT_ENTRIES] =
    [core::ptr::null(); MAX_MADT_ENTRIES];
static mut MADT_NMIS: [*const MadtNmi; MAX_MADT_ENTRIES] = [core::ptr::null(); MAX_MADT_ENTRIES];

pub static MADT_LOCAL_APIC_COUNT: AtomicU64 = AtomicU64::new(0);
static MADT_NMI_COUNT: AtomicU64 = AtomicU64::new(0);

static X2APIC: AtomicBool = AtomicBool::new(false);
pub static BSP_LOCAL_APIC_ID: AtomicU64 = AtomicU64::new(0);
static LAPIC_ADDR: AtomicU64 = AtomicU64::new(0);

pub fn init() {
    MADT_PTR.store(
        acpi::find_sdt(b"APIC", 0).expect("Did not find APIC table?") as u64,
        Ordering::Relaxed,
    );

    let madt_ptr = MADT_PTR.load(Ordering::Relaxed) as *const Madt;
    let madt_len = unsafe { (*madt_ptr).header.length } as u64;
    let entry0_addr = unsafe { &(*madt_ptr).entry0 as *const MadtHeader as u64 };
    let madt_end = madt_ptr as u64 + madt_len;

    let mut local_apic_count = 0u64;
    let mut nmi_count = 0u64;

    let mut cursor = entry0_addr;
    while cursor < madt_end {
        let entry = unsafe { &*(cursor as *const MadtHeader) };
        match entry.entry_type {
            MADT_TYPE_PROCESSOR_LOCAL_APIC => {
                unsafe {
                    MADT_LOCAL_APICS[local_apic_count as usize] =
                        cursor as *const MadtProcessorLocalApic;
                }
                local_apic_count += 1;
            }
            MADT_TYPE_NON_MASKABLE_INTERRUPT => {
                unsafe {
                    MADT_NMIS[nmi_count as usize] = cursor as *const MadtNmi;
                }
                nmi_count += 1;
            }
            _ => {}
        }
        cursor += entry.record_length as u64;
    }

    MADT_LOCAL_APIC_COUNT.store(local_apic_count, Ordering::Relaxed);
    MADT_NMI_COUNT.store(nmi_count, Ordering::Relaxed);

    LAPIC_ADDR.store(
        0xfee00000 + HHDM_OFFSET.load(Ordering::Relaxed),
        Ordering::Relaxed,
    );

    let cpuid1 = cpuid(1, 0);
    if cpuid1.ecx & (1 << 21) != 0 {
        X2APIC.store(true, Ordering::Relaxed);
    }

    let madt_flags = unsafe { (*madt_ptr).flags };
    if madt_flags & 1 != 0 {
        // Remap the legacy PIC, then mask it off entirely.
        unsafe {
            outb(0x20, 0x11);
            outb(0xA0, 0x11);
            outb(0x21, 0x20);
            outb(0xA1, 0x28);
            outb(0x21, 4);
            outb(0xA1, 2);
            outb(0x21, 1);
            outb(0xA1, 1);
            outb(0x21, 0);
            outb(0xA1, 0);

            // Disable the PIC
            outb(0xA1, 0xFF);
            outb(0x21, 0xFF);
        }
    }
}

fn reg_to_x2apic(reg: u32) -> u32 {
    let x2apic_reg = if reg == 0x310 { 0x30 } else { reg >> 4 };
    x2apic_reg + 0x800
}

pub struct Lapic;

impl Lapic {
    pub fn read(reg: u32) -> u32 {
        if X2APIC.load(Ordering::Relaxed) {
            unsafe { rdmsr(reg_to_x2apic(reg)) as u32 }
        } else {
            let addr = LAPIC_ADDR.load(Ordering::Relaxed) + reg as u64;
            unsafe { core::ptr::read_volatile(addr as *const u32) }
        }
    }

    pub fn write(reg: u32, value: u32) {
        if X2APIC.load(Ordering::Relaxed) {
            unsafe { wrmsr(reg_to_x2apic(reg), value as u64) };
        } else {
            let addr = LAPIC_ADDR.load(Ordering::Relaxed) + reg as u64;
            unsafe { core::ptr::write_volatile(addr as *mut u32, value) };
        }
    }

    pub fn get_id() -> u8 {
        (Self::read(0x20) >> 24) as u8
    }

    fn set_nmi(vec: u8, current_processor_id: u8, processor_id: u8, flags: u16, lint: u8) {
        if processor_id != 0xFF && current_processor_id != processor_id {
            return;
        }

        let mut nmi: u32 = 0x400 | vec as u32;
        if flags & 2 != 0 {
            nmi |= 1 << 13; // active low
        }
        if flags & 8 != 0 {
            nmi |= 1 << 15; // level triggered
        }

        match lint {
            0 => Self::write(0x350, nmi),
            1 => Self::write(0x360, nmi),
            _ => {}
        }
    }

    pub fn eoi() {
        Self::write(0xB0, 0);
    }

    pub fn init(processor_id: u8) {
        unsafe {
            wrmsr(0x1b, (1 << 11) | rdmsr(0x1b));
            if X2APIC.load(Ordering::Relaxed) {
                wrmsr(0x1b, (1 << 10) | rdmsr(0x1b));
            }
        }

        Self::write(0x80, 0);
        Self::write(0xF0, Self::read(0xF0) | 0x100);

        if !X2APIC.load(Ordering::Relaxed) {
            Self::write(0xE0, 0xF0000000);
            Self::write(0xD0, Self::read(0x20));
        }

        let nmi_count = MADT_NMI_COUNT.load(Ordering::Relaxed);
        for i in 0..nmi_count {
            let nmi = unsafe { &*MADT_NMIS[i as usize] };
            Self::set_nmi(2, processor_id, nmi.acpi_processor_id, nmi.flags, nmi.lint);
        }
    }

    pub fn send_ipi(lapic_id: u32, flags: u32) {
        if X2APIC.load(Ordering::Relaxed) {
            unsafe { wrmsr(0x830, ((lapic_id as u64) << 32) | flags as u64) };
        } else {
            Self::write(0x310, lapic_id << 24);
            Self::write(0x300, flags);
        }
    }

    pub fn send_ipi_to(processor: u32, vector: u32) {
        Self::send_ipi(processor, vector);
    }

    pub fn broadcast_ipi(vector: u32, include_me: bool) {
        if include_me {
            Self::send_ipi(0, (2 << 18) | vector);
        } else {
            Self::send_ipi(0, (3 << 18) | vector);
        }
    }
}
